File: apps/web/src/shared/api/friends.ts
import type { PresenceStatus } from "./presence";
import { apiRequest } from "./client";

export type FriendshipState =
  | "self"
  | "friend"
  | "incoming_request"
  | "outgoing_request"
  | "none";

export type FriendSummary = {
  friendship_id: string;
  user_id: string;
  username: string;
  friends_since: string;
  presence_status: PresenceStatus;
};

export type FriendRequestSummary = {
  id: string;
  requester_user_id: string;
  requester_username: string;
  recipient_user_id: string;
  recipient_username: string;
  request_text: string | null;
  status: "pending" | "accepted" | "rejected" | "cancelled";
  created_at: string;
};

export type FriendListResponse = {
  friends: FriendSummary[];
};

export type FriendRequestListResponse = {
  incoming_requests: FriendRequestSummary[];
  outgoing_requests: FriendRequestSummary[];
};

export type CreateFriendRequestPayload = {
  username: string;
  message?: string;
};

export const friendsApi = {
  listFriends: () => apiRequest<FriendListResponse>("/api/friends"),
  listRequests: () => apiRequest<FriendRequestListResponse>("/api/friends/requests"),
  sendRequest: (payload: CreateFriendRequestPayload) =>
    apiRequest<FriendRequestSummary>("/api/friends/requests", {
      method: "POST",
      body: JSON.stringify(payload),
    }),
  acceptRequest: (requestId: string) =>
    apiRequest<FriendSummary>(`/api/friends/requests/${requestId}/accept`, {
      method: "POST",
    }),
  rejectRequest: (requestId: string) =>
    apiRequest<{ success: boolean; message: string }>(`/api/friends/requests/${requestId}/reject`, {
      method: "POST",
    }),
  removeFriend: (friendUserId: string) =>
    apiRequest<{ success: boolean; message: string }>(`/api/friends/${friendUserId}`, {
      method: "DELETE",
    }),
};

File: apps/api/app/api/schemas/friends.py
from __future__ import annotations

from datetime import datetime
from typing import Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, StringConstraints

from app.api.schemas.presence import PresenceStatus
from app.db.models.enums import FriendRequestStatus

FriendshipState = Literal["self", "friend", "incoming_request", "outgoing_request", "none"]


class FriendSummaryResponse(BaseModel):
    friendship_id: UUID
    user_id: UUID
    username: str
    friends_since: datetime
    presence_status: PresenceStatus = "offline"


class FriendListResponse(BaseModel):
    friends: list[FriendSummaryResponse]


class FriendRequestSummaryResponse(BaseModel):
    id: UUID
    requester_user_id: UUID
    requester_username: str
    recipient_user_id: UUID
    recipient_username: str
    request_text: str | None
    status: FriendRequestStatus
    created_at: datetime


class FriendRequestListResponse(BaseModel):
    incoming_requests: list[FriendRequestSummaryResponse]
    outgoing_requests: list[FriendRequestSummaryResponse]


class CreateFriendRequestRequest(BaseModel):
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "username": "alice",
                "message": "Want to connect for project collaboration?",
            }
        }
    )

    username: Annotated[
        str,
        StringConstraints(
            strip_whitespace=True,
            min_length=3,
            max_length=64,
            pattern=r"^[A-Za-z0-9_.-]+$",
        ),
    ]
    message: (
        Annotated[
            str,
            StringConstraints(strip_whitespace=True, min_length=1, max_length=500),
        ]
        | None
    ) = None
